test-data/unit/check-overloading.test

```diff
@@ -1136,3 +1136,24 @@ def f(a: Callable[[], int]) -> None: pass
 @overload
 def f(a: str) -> None: pass
 f(0)  # E: No overload variant of "f" matches argument types [builtins.int]
+
+[case testCustomRedefinitionDecorator]
+from typing import Any, Callable, Type
+
+class Chain(object):
+    def chain(self, function: Callable[[Any], int]) -> 'Chain':
+        return self
+
+class Test(object):
+    do_chain = Chain()
+
+    @do_chain.chain
+    def do_chain(self) -> int:
+        return 2
+
+    @do_chain.chain  # E: Name 'do_chain' already defined
```

Comment:
I'm not sure what constitutes a "good" redefinition as distinct from a "bad" redefinition that we want to error on.  Right now we specifically treat redefinitions with decorators that refer to properties and overloads specially, and think every other redefinition is an error.

I don't know why the first `def` here isn't a redefinition error offhand; I'll read through the code in more detail tomorrow I think. If you need a tighter revision cycle than that on this PR, feel free to steal or preempt it -- I'm busy today from noon on, and won't be back to my computer until tomorrow late afternoon or early evening.